verification/spec/connection-handshake/Environment.tla

```diff
@@ -0,0 +1,402 @@
+          /\ inBufChainB = <<>>
+       \/ /\ inBufChainB \in {<<msg>> : (* ICS3MsgInit to chain B. *)
+            msg \in InitMsgs(ChainBConnectionEnds, ChainAConnectionEnds)}
+          /\ inBufChainB = <<>>
+       \/ /\ inBufChainA \in {<<msg>> : (* ICS3MsgInit to both chains. *)
+            msg \in InitMsgs(ChainAConnectionEnds, ChainBConnectionEnds)}
+          /\ inBufChainB \in {<<msg>> :
+            msg \in InitMsgs(ChainBConnectionEnds, ChainAConnectionEnds)}
+    /\ outBufChainA = <<>>
+    /\ outBufChainB = <<>>
+
+
+Relay(from, to) ==
+    /\ from # <<>>
+    /\ Len(to) < MaxBufLen - 1
+    /\ to' = Append(to, Head(from))
+    /\ from' = Tail(from)
+
+
+(* Default next (good) action for Environment.
+
+    TODO: should advance the height non-deterministically?
+
+    May change either of the store of chain A or B. 
+ *)
+GoodNextEnv ==
+    \/ /\ chmA!CanAdvance
+       /\ \/ chmA!AdvanceChainHeight
+          \/ chmA!UpdateClient(storeChainB.latestHeight)
+       /\ UNCHANGED<<storeChainB, outBufChainA, outBufChainB, inBufChainA, inBufChainB>>
+    \/ /\ chmB!CanAdvance
+       /\ \/ chmB!AdvanceChainHeight
+          \/ chmB!UpdateClient(storeChainA.latestHeight)
+       /\ UNCHANGED<<storeChainA, outBufChainA, outBufChainB, inBufChainA, inBufChainB>>
+
+
+RelayNextEnv ==
+    \/ LET msg == Head(outBufChainA)
+           targetHeight == IF MessageTypeIncludesConnProof(msg.type)
+                           THEN msg.connProof.height
+                           ELSE storeChainA.latestHeight
+       IN /\ Relay(outBufChainA, inBufChainB)
+          /\ \/ chmB!CanAdvance /\ chmB!UpdateClient(targetHeight)
+             \/ ~ chmB!CanAdvance /\ UNCHANGED storeChainB
+          /\ UNCHANGED<<storeChainA, outBufChainB, inBufChainA>>
+    \/ LET msg == Head(outBufChainB)
+           targetHeight == IF MessageTypeIncludesConnProof(msg.type)
+                           THEN msg.connProof.height
+                           ELSE storeChainB.latestHeight
+       IN /\ Relay(outBufChainB, inBufChainA)
+          /\ \/ chmA!CanAdvance /\ chmA!UpdateClient(targetHeight)
+             \/ ~ chmA!CanAdvance /\ UNCHANGED storeChainA
+          /\ UNCHANGED<<storeChainB, outBufChainA, inBufChainB>>
+
+
+(* Environment malicious behavior.
+
+    The environment injects a msg. in the buffer of one of the chains. 
+    This interferes with the ICS3 protocol by introducing additional
+    messages that are usually incorrect.
+    
+    Without the first constraint, on the "Len(inBufChainA)" and
+    "Len(inBufChainB)", Env could fill buffers (DoS attack). This can
+    lead to a deadlock, because chains will simply be unable to reply
+    to each other.
+ *)
+MaliciousNextEnv ==
+    \/ /\ Len(inBufChainA) < MaxBufLen - 1
+       /\ inBufChainA' \in {Append(inBufChainA, arbitraryMsg) : 
+            arbitraryMsg \in {msg \in ConnectionHandshakeMessages : 
+            msg.type = "ICS3MsgTry"}}
```

Comment:
Why msg type is constrained here? We want to allow all messages to be created by adversary.

Reply:
I added this artificial constraint on "ICS3MsgTry" to make the checking more tractable. But even with this constraint, after 1 hour of execution (~50 milion distinct states), TLC didn't finish on my laptop.

I am removing this constraint.